### ISSUE TITLE
Make it possible to complete disable the ImageJ updater

### ISSUE THREAD

**Issue by ctrueden** [imagej/imagej-updater#7]
Sometimes people want to deploy a version of Fiji that never checks for updates, and/or does not even have updater functionality at all, preferring to carefully control the updates themselves using an alternative mechanism (e.g., Git).

We should make sure the architecture of ImageJ2 allows for this.

For example, right now, Fiji tries to invoke the remote ImageJ updater if ij-updater*.jar is missing from its installation, and ImageJ2 has a hardcoded invocation of the `UpToDate` command directly in `AbstractUserInterface`, which is definitely the wrong approach. Better would be for commands to be able to annotate themselves as auto-run, and/or have a flexible auto-run mechanism that by default ships with the UpToDate check included. That way, users can just remove the UpToDate check from the auto-run configuration to disable the updater.

Migrated-From: http://trac.imagej.net/ticket/2035

**Comment by ctrueden** [comment]
For those stumbling across this issue these days: note that `Fiji_Updater` is now obsolete, and removing it will _not_ disable the updater. This is because `Fiji_Updater` is simply a backwards compatibility stub that adds the "Update Fiji" command to the help menu. The actual updater now lives in imagej-ui-swing. So it no longer possible to disable the updater by deleting a particular JAR file.

However, here is one possible way of disabling it. Create the following class, make a JAR and put that JAR into the `ImageJ.app/jars` folder:

```java

package net.imagej.updater;

import org.scijava.Priority;
import org.scijava.plugin.Plugin;
import org.scijava.service.Service;
import org.scijava.ui.event.UIShownEvent;

/** {@link UpdateService} which disables update checking at startup. */
@Plugin(type = Service.class, priority = Priority.HIGH_PRIORITY)
public class NoUpdaterUpdateService extends DefaultUpdateService {

	// -- Event handlers --

	/** Override the initial update check, to never do anything. */
	@Override
	protected void onEvent(final UIShownEvent evt) {
		// NB: Do nothing.
	}

}
```

**Comment by eCrofey** [comment]
Hi i i'm a pacakger and need to dissable the update function,but my coding skills 're very small.
can you please update the files so I can intergrate it into my msi pacakge I create.
it would help me a lot.

**Comment by ctrueden** [comment]
@eCrofey You can simply delete `jars/imagej-updater-*.jar`, also optionally `jars/imagej-plugins-uploader-*.jar`.

**Comment by eCrofey** [comment]
Great thanx a Lot !

**Comment by imagejan** [comment]
Reopening this, as it seems there are some community use cases (packaging, centralized management of institutional installations, ...) where we need a better (and simple) way to disable the updater.

**Comment by ctrueden** [comment]
As of imagej/imagej-updater@b69709ca8a0af98f8331b8ce9ba19731a4f90b73, part of imagej-updater-0.10.4 and later, you can pass `-Dimagej.updater.disableAutocheck=true` when launching, and the automatic update check will be skipped.

What else does the community need here? Do we need an environment variable? Maybe something like `IMAGEJ_UPDATER_DISABLE=1` which makes the Updater _always_ refuse to run? Do we also need a system property to similarly disable the Updater *completely* rather than only the auto-update check at the beginning?

@lguerard @eCrofey Your input here would be helpful. What do you want to have happen when the user tries to run _Help > Update..._? Nothing? An error?

**Comment by lguerard** [comment]
Thanks for the info, didn't know about the launch command !

> What else does the community need here? Do we need an environment variable? Maybe something like IMAGEJ_UPDATER_DISABLE=1 which makes the Updater always refuse to run? Do we also need a system property to similarly disable the Updater completely rather than only the auto-update check at the beginning?

The main issue we have right now, is that removing the updater JAR breaks everything deep learning in Fiji. I guess disabling the updater would do the same no ? 
If it doesn't, then this would be perfect. But If it does, then I guess just disabling the auto-update check at the beginning should be enough as (hopefully) users won't be tempted to update unless they really need/want it. What do you think @ehrenfeu ?

**Comment by ctrueden** [comment]
> I guess disabling the updater would do the same no ?

I don't think so, because the `UpdateService` and related classes would still be there and available, they just wouldn't do anything.

What I need to know is: how are you deploying this to users? Do you have control over:
* the program launch (i.e. you could add `-Dimagej.updater.disableAutocheck=true`)
* the environment (i.e. you could set `IMAGEJ_UPDATER_DISABLE=1`)
* the packaging (e.g. you could tweak something in the .zip users download such as adding an `imagej2.cfg` file with an `[updater]` section with `enabled = false`)
* something else?

**Comment by lguerard** [comment]
> What I need to know is: how are you deploying this to users?

The way we do it is to download a brand new Fiji, add updates sites with a python script and then run the updater. Once this is done we move the updater JAR to a different a `disabled/jars/` folder. Then we copy the whole Fiji folder to all our machines and put the shortcut on the desktop. We also add an extra `imageJ.cfg` file to set up the RAM usage based on specific machines we have. The deployment to all the machines as well as the shortcut creation is done through a PowerShell script.

So to answer your questions, we control more or less everything inside the Fiji folder and shortcut. I could have a look at adding the launch argument to the shortcut as well and see if that works.

**Comment by DuncanClarke** [comment]
Sorry to drag up this old query; just wanted to know if there is an official deployment method which doesn't allow users to run updates. I mean we can add a link with the -Dimagej.updater.disableAutocheck=true but users being users, I suspect they would find a way to launch it without this. 

What we are looking for is a way to install this system-wide (C:\Program Files) such that users won't get errors from attempting to run updates. If an update is needed, we will do it centrally. Any "personal" installs will be blocked from running.

**Comment by mkitti** [comment]
I'm going to guess this would be more easily addressed with https://github.com/scijava/jaunch
